r/tests/testthat/test-dplyr-join.R

```diff
@@ -0,0 +1,108 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+skip_if_not_available("dataset")
+
+library(dplyr)
+
+left <- example_data
+# Error: Invalid: Dictionary type support for join output field is not yet implemented, output field reference: FieldRef.Name(fct) on left side of the join
+# (select(-fct) also solves this but remove once)
+left$fct <- NULL
+left$some_grouping <- rep(c(1, 2), 5)
+
+to_join <- tibble::tibble(
+  some_grouping = c(1, 2),
+  capital_letters = c("A", "B"),
+  another_column = TRUE
+)
+
+test_that("left_join", {
+  expect_dplyr_equal(
+    input %>%
+      left_join(to_join) %>%
+      collect(),
+    left
+  )
+})
+
+test_that("left_join `by` args", {
+  expect_dplyr_equal(
+    input %>%
+      left_join(to_join, by = "some_grouping") %>%
+      collect(),
+    left
+  )
+  expect_dplyr_equal(
+    input %>%
+      left_join(
+        to_join %>%
+          rename(the_grouping = some_grouping),
+        by = c(some_grouping = "the_grouping")
+      ) %>%
+      collect(),
+    left
+  )
+})
+
+# TODO: test duplicate col names
+# TODO: test invalid 'by'
+
+test_that("right_join", {
+  expect_dplyr_equal(
+    input %>%
+      right_join(to_join) %>%
+      collect(),
+    left
+  )
+})
+
+test_that("inner_join", {
+  expect_dplyr_equal(
+    input %>%
+      inner_join(to_join) %>%
+      collect(),
+    left
+  )
+})
+
+test_that("full_join", {
+  expect_dplyr_equal(
+    input %>%
+      full_join(to_join) %>%
+      collect(),
+    left
+  )
+})
+
+test_that("semi_join", {
+  expect_dplyr_equal(
+    input %>%
+      semi_join(to_join) %>%
+      collect(),
+    left
+  )
+})
+
+test_that("anti_join", {
+  expect_dplyr_equal(
+    input %>%
+      anti_join(to_join) %>%
+      collect(),
+    left
+  )
+})
```

Comment:
I can do this when I rebase, but should we have tests for the error handling as well (e.g. columns named but not in `x` or `y`, etc.)?

Reply:
Oh yeah, there are lots of missing tests, and these datasets aren't ideal for testing joins. We may want to just use the datasets that dplyr uses in its tests or something.